### DISCUSSION TITLE
App layout

### DISCUSSION THREAD

**Discussion by zhuiks** [ezra-bible-app/ezra-bible-app#299] [Development]
Currently, we have this kind of layout (with future interlinear view):
![Current layout](https://user-images.githubusercontent.com/3584119/118908280-be528500-b8e6-11eb-8686-6a4fd9d828c4.png)

This not reflects nice in portrait view (not even bringing up mobile devices):
![Current layout(1)](https://user-images.githubusercontent.com/3584119/118908375-f1951400-b8e6-11eb-84c1-2138cbbad1f3.png)

What if we combine some views:
![Future layout](https://user-images.githubusercontent.com/3584119/118908498-2903c080-b8e7-11eb-96cf-47309f5fe703.jpg)

Bottom view can be shared between Dictionary and Interlinear and can be hidden
Left sidebar currently will be only for tags but in future can be shared with notes with wiki functionality #137 and we can even move chapter navigation/book selection to this area.

![Future layout(1)](https://user-images.githubusercontent.com/3584119/118909148-32415d00-b8e8-11eb-819f-0d50f185dc57.jpg)

On smaller portrait left sidebar can be pan overlay (a common design pattern).

I'm just thinking on how properly implement #77...

**Comment by tobias-klein** [comment]
Thanks for this proposal! In the original version of Ezra, I actually had the tag list and a book outline combined in the left sidebar. Yes, I suppose we can think about putting the book navigation there. This may be especially useful when you have turned on the section header navigation. In my original version I implemented this as tabs for the left sidebar, so you could switch between the tag list and the book's outline. How do you imagine the details for the left sidebar? Using tabs? Or something else?

Combining the dictionary and the (future) interlinear view also makes sense.

Note that when I am using Ezra in portrait mode on my tablet, the layout is currently configured in a way where the left sidebar is not used at all and the tags list rather moves to the dropdown via "change tags". That works pretty well.

**Reply by zhuiks** [reply]
> How do you imagine the details for the left sidebar? Using tabs? Or something else?

Seems like it's quite common to have icon-buttons on the left side:
VSCode:
![Screenshot from 2021-05-21 19-18-15](https://user-images.githubusercontent.com/3584119/119208805-59736800-ba69-11eb-836f-637a285fcbf7.png)

Opera:
![Screenshot from 2021-05-21 19-12-06](https://user-images.githubusercontent.com/3584119/119208845-77d96380-ba69-11eb-915e-307820c64759.png)

**Reply by zhuiks** [reply]
I also like the idea of settings panel in Opera instead of popup dialog:
![Screenshot from 2021-05-21 19-12-30](https://user-images.githubusercontent.com/3584119/119208985-06e67b80-ba6a-11eb-9d6b-678b03edfc9b.png)

**Comment by zhuiks** [comment]
So, continuing conversation about layout from #302:

>   * We should keep the option of having the dictionary below the tag list - from a formatting perspective this makes more sense because the dictionary tends to need more vertical space than horizontal space, especially if you have multiple dictionaries installed.
> 
>   * So I think the dictionary should be below the tag list in the side bar by default and if there is no sidebar then it can be shown below the Bible browser.

Here are some sketches of the layout:
![Frame 1(2)](https://user-images.githubusercontent.com/3584119/119510083-dab34f00-bd36-11eb-90aa-935ef03cc0c6.png)

Or in portrait mode:
![Frame 2(3)](https://user-images.githubusercontent.com/3584119/119510159-e99a0180-bd36-11eb-9dde-01d1245abdfd.png)



* Area **2** is always the Bible browser with tabs and always visible.
* Area **1** currently is the tag list, But in future we can move chapter/headers navigation (and maybe book selection - if there is a nice UX solution) there and notes list. Those can be switchable between each other. I.e. only one of them visible at the time. The area itself should be collapsible. I think the proper name for it is a *sidebar*
* Area **4** seems to be perfect for interlinear details. It can be shared with dictionary. It should be collapsible. Is it *bottom bar* or *detail view*?
* Area **3**... Is it a part of 1 or a part of 4? Or is it a separate thing? If it is a separate thing, then it can be a good place for a dictionary (as kind of it is now). Obviously when only 2 and 3 are visible the look should be the same as 2 and 4.

### Thoughts:
* Layout should be done in pure CSS, fully responsive including smaller screens and screen rotations! Use JS only for switching views visibility (probably with `display: none`). I think adding classes via JS and styling them with CSS is an ambiguous work around way. That was the only way before. But nowadays CSS is very powerful.
* On smaller (mobile) screens in portrait mode area 1 can be displayed as overlay, instead of pushing area 2 to the right. It's a common pattern with mobile apps, at least with Android.
* As noted, dictionary layout is more vertical (needs more vertical space than horizontal). As the solution to displaying dictionary in horizontal view (i.e area 4) it can have multiple columns for several dictionary articles

### Questions:
* Can interlinear and dictionary views be visible at the same time. Would it be helpful or just duplicated information?
* Any other possible additional views to be implemented in the far future?
* Should area 3 be the separate area or is it overloading UI?

**Reply by tobias-klein** [reply]
Regarding your questions:
Yes, I imagine that it could be useful to have interlinear view and dictionary view visible at the same time. They have different purposes, one shows different dictionary entries for one particular word, the other shows the original language parallel to grammar details and parallel translated words (but without word definitions from a dictionary).

I would see the dictionary as part of the side bar as long as the side bar is existing (like right now). But if there is no sidebar we would need to move it into area 4.

**Reply by tobias-klein** [reply]
Any other possible views ... well, the only thing remotely on our agenda is the wiki functionality and that's not fully designed yet.

**Comment by zhuiks** [comment]
Here is another sketch (work in progress):
![Future layout(5)](https://user-images.githubusercontent.com/3584119/119569359-44058300-bd74-11eb-8cfb-dd57b602a28f.jpg)

---
Updated the sketch.

It's a bit ugly, but it's more an idea.
**Dictionary** definitions from different modules can be displayed as "cards" to layout nicely in horizontal view.

**Reply by tobias-klein** [reply]
I like the idea of the icon bar (VS Code Style).
The dictionary mock up looks nice, too. So with that proposal we could have a responsive layout depending on whether this is rendered vertically in the sidebar or horizontally under the Bible browser.

**Comment by tobias-klein** [comment]
Let's discuss in detail in our next Skype call! I'll try to get my head around your proposals / thoughts until then :)

**Reply by zhuiks** [reply]
I updated few sketches. Hope this adds some clarity to the root of the problem

**Comment by tobias-klein** [comment]
@augustin-colesnic Feel free to get involved with the discussion going on here!

**Comment by zhuiks** [comment]
While implementing #526 I realized that the layout can be simplified.
It's important to see Bible browser while along with dictionary or tags toolbox. But I don't think it's crucial to see tags and dictionary together. It just adds extra visual noise for the user.

So for desktop mode those can be in side panel:

![Ezra 22-Nov-2021-195107](https://user-images.githubusercontent.com/3584119/142962273-43a8a34f-982f-4ca7-a094-557d8ad78340.gif)

But for portrait/tablet mode, tool windows can be at the bottom (one at a time):
![Ezra 22-Nov-2021-200823](https://user-images.githubusercontent.com/3584119/142962281-9b68e098-46f7-4b43-827c-8fc0085c85ce.gif)

@tobias-klein @augustin-colesnic @debritto what do you think?

**Reply by debritto** [reply]
Hi @zhuiks ,
Good point. I think you're right, at least for me it's not crucial to see tags and dictionary together. Showing just one at a time will make the UI cleaner. :)

**Reply by tobias-klein** [reply]
@zhuiks 
I like this  idea! Yes, it's probably OK to assume that you're not working with multiple toolbox panels at once. Therefore this is a good way of simplifying the UI. Feel free to go for it! :)

**Reply by augustin-colesnic** [reply]
Hey  @zhuiks, nice improvement to the UI!
